### PR TITLE
Elektrify lcdproc with low-level implementation

### PR DESCRIPTION
**Pull request by kodebach**
This pull request, contains a low-level libelektra implementation of lcdproc.

Whether or not we want to merge this into master is up to you @markus2330.
I mostly created this pull request, so you could easily find my changes and comment on them (you probably should merge #3 first).

### CODE REVIEW

File: readconf.py [hunk 1]
```diff
@@ -0,0 +1,193 @@
+#!/bin/python3
```
Comment (markus2330) [inline readconf.py]: env python3

File: readconf.py [hunk 2]
```diff
@@ -0,0 +1,193 @@
...
+            kdb_config[kdb_key] = config[section][key]
+
+    return kdb_config
+
+
+def write_to_kdb(kdb_config):
+    def write_value(key, value):
+        if value.startswith('"') and value.endswith('"'):
+            value = value[1:-1]
+        cmd = ["kdb", "set", "--", f"{key}", f"{value}"]
```
Comment (markus2330) [inline readconf.py:161]: why not using the python binding?

File: include/elektra/kdb.h [hunk 3]
```diff
@@ -0,0 +1,252 @@
+/**
```
Comment (markus2330) [inline include/elektra/kdb.h]: This file should be removed.

File: LCDd.conf [hunk 4]
```diff
@@ -530,10 +530,10 @@ PixelShiftY=2
 # select the serial device to use [default: /dev/lcd]
 Device=/dev/lcd
 
-# set the initial contrast value [default: 560; legal: 0 - 1000]
-Contrast=560
+# set the initial contrast value [default: 500; legal: 0 - 1000]
+Contrast=500
```
Comment (markus2330) [inline LCDd.conf:534]: Are these bugs you found? These should definitely go upstream.

File: clients/lcdexec/lcdexec.c [hunk 5]
```diff
@@ -42,6 +42,10 @@
 #define DEFAULT_CONFIGFILE	SYSCONFDIR "/lcdexec.conf"
 #define DEFAULT_PIDFILE		PIDFILEDIR "/lcdexec.pid"
 
+#ifdef HAVE_CONFIG_H
+#include "config.h"
```
Comment (markus2330) [inline clients/lcdexec/lcdexec.c:46]: Why is this needed?

File: clients/lcdexec/menu.c [hunk 6]
```diff
@@ -28,222 +28,396 @@
...
+		if(strcasecmp(paramTypes[i], type) == 0) {
+			typeId = i;
+			break;
 		}
+	}
+	free(dupName);
+	
+	char* keyName;
+	char* keyBaseNameEnd;
+	switch(typeId) {
```
Comment (markus2330) [inline clients/lcdexec/menu.c]: Please document code a little, what is happening here?

File: server/drivers/serialPOS_cd5220.c [hunk 7]
```diff
@@ -0,0 +1,363 @@
+/**
```
Comment (markus2330) [inline server/drivers/serialPOS_cd5220.c:1]: Why are these files new?

File: shared/elektraconfig.c [hunk 8]
```diff
@@ -0,0 +1,193 @@
...
+		keyDel(key);
+		return default_value;
+	}
+
+	const char* str = keyString(key);
+	keyDel(key);
+
+	// TODO (kodebach): use boolean plugin to normalize values to 0/1
+	if ((strcasecmp(str, "0") == 0) || (strcasecmp(str, "false") == 0) ||
+	    (strcasecmp(str, "n") == 0) || (strcasecmp(str, "no") == 0) ||
```
Comment (markus2330) [inline shared/elektraconfig.c:96]: Is this compatible to what the boolean plugin supports?

File: shared/elektraconfig.c [hunk 9]
```diff
@@ -0,0 +1,193 @@
...
+    }
+
+	return default_value;
+}
+
+bool econfig_get_bool(Config* config, const char* keyName, const bool default_value)
+{
+	Key* key = ksLookupByName(config, keyName, 0);
+	if (key == NULL) {
+		keyDel(key);
```
Comment (markus2330) [inline shared/elektraconfig.c]: keyDel(NULL)?

**Review by markus2330 — CHANGES_REQUESTED** [review]
Thank you, great work!

**Comment by markus2330** [comment]
If you can build on this, we can merge it to master, otherwise we should use a different branch.

File: readconf.py (hunk 2 above)
Comment (kodebach) [reply readconf.py:161]: this was simply quicker than looking into the python binding

File: clients/lcdexec/lcdexec.c (hunk 5 above)
Comment (kodebach) [reply clients/lcdexec/lcdexec.c:46]: this comes from autoconf
the `config.h` file contains some defines that are used here e.g. `HAVE_SA_RESTART`

this is probably also something that should go upstream

File: server/drivers/serialPOS_cd5220.c (hunk 7 above)
Comment (kodebach) [reply server/drivers/serialPOS_cd5220.c:1]: good question... the file is identical to the one on the [master branch](https://github.com/ElektraInitiative/lcdproc/blob/master/server/drivers/serialPOS_cd5220.c)

File: shared/elektraconfig.c (hunk 8 above)
Comment (kodebach) [reply shared/elektraconfig.c:96]: it is not the default but the plugin can certainly be configured this way

File: server/drivers/serialPOS_cd5220.c (hunk 7 above)
Comment (markus2330) [reply server/drivers/serialPOS_cd5220.c:1]: Might be a bug of GitHub. Maybe a rebase or recreation of the PR is needed.

File: shared/elektraconfig.c (hunk 8 above)
Comment (markus2330) [reply shared/elektraconfig.c:96]: Yes, would be good if we have all of such code in plugins. Then one can easily change to other plugins without having any hard-coded, unwanted behavior. (For example, some people might want to have stricter or localized booleans)

File: shared/elektraconfig.c (hunk 9 above)
Comment (markus2330) [reply shared/elektraconfig.c]: Still open?

File: readconf.py (hunk 2 above)
Comment (markus2330) [reply readconf.py:161]: I know what you mean. But the python binding is quite easy to use and already has docu.

We maybe do not need the readconf.py anyway, so there might be no need to fix this.

File: server/drivers/serialPOS_cd5220.c (hunk 7 above)
Comment (kodebach) [reply server/drivers/serialPOS_cd5220.c:1]: Once you are happy with the code, I think I will just push it directly to a separate branch. Just in case this PR is actually broken somehow.

Comment (markus2330) [reply server/drivers/serialPOS_cd5220.c:1]: Yes, you can already push it to a separate branch.

I think you should continue with the high-level API. Your benchmarks are the only customer of the low-level API anyway.

**Comment by kodebach** [comment]
Pushed to elektra_lowlevel -> closing
